share/eds.go

```diff
@@ -0,0 +1,215 @@
+	// shares are extracted from the eds so that we can reimport them to traverse
+	shares := ipld.ExtractEDS(eds)
+	if len(shares) == 0 {
+		return nil, fmt.Errorf("share: importing empty data")
+	}
+	// todo: add correct batch size here
+	squareSize := int(math.Sqrt(float64(len(shares))))
+	batchAdder := ipld.NewNmtNodeAdder(ctx, bs, format.MaxSizeBatchOption(squareSize/2))
+	// this adder ignores leaves, so that they are not added to the store we iterate through in writeProofs
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(squareSize/2), nmt.NodeVisitor(batchAdder.VisitInnerNodes))
+	eds, err := rsmt2d.ImportExtendedDataSquare(shares, ipld.DefaultRSMT2DCodec(), tree.Constructor)
+	if err != nil {
+		return nil, fmt.Errorf("failure to recompute the extended data square: %w", err)
+	}
+	// compute roots
+	eds.RowRoots()
+	// commit the batch to DAG
+	err = batchAdder.Commit()
+	if err != nil {
+		return nil, fmt.Errorf("failure to commit the inner nodes to the dag: %w", err)
+	}
+
+	return &writingSession{
+		eds:   eds,
+		store: store,
+		w:     w,
+	}, nil
+}
+
+// writeHeader creates a CarV1 header using the EDS's Row and Column roots as the list of DAG roots.
+func (w *writingSession) writeHeader() error {
+	rootCids, err := rootsToCids(w.eds)
+	if err != nil {
+		return fmt.Errorf("failure getting root cids: %w", err)
+	}
+
+	err = car.WriteHeader(&car.CarHeader{
+		Roots:   rootCids,
+		Version: 1,
+	}, w.w)
+	if err != nil {
+		return fmt.Errorf("failure writing carv1 header: %w", err)
+	}
+	return nil
+}
+
+// writeQuadrants reorders the shares to quadrant order and writes them to the CARv1 file.
+func (w *writingSession) writeQuadrants() error {
+	shares := quadrantOrder(w.eds)
+	for _, share := range shares {
+		cid, err := plugin.CidFromNamespacedSha256(nmt.Sha256Namespace8FlaggedLeaf(share))
+		if err != nil {
+			return fmt.Errorf("failure to get cid from share: %w", err)
+		}
+		err = util.LdWrite(w.w, cid.Bytes(), share)
+		if err != nil {
+			return fmt.Errorf("failure to write share: %w", err)
+		}
+	}
+	return nil
+}
+
+// writeProofs iterates over the in-memory blockstore's keys and writes all inner nodes to the CARv1 file.
+func (w *writingSession) writeProofs(ctx context.Context) error {
+	// we only stored proofs to the store, so we can just iterate over them here without getting any leaves
+	proofs, err := w.store.AllKeysChan(ctx)
+	if err != nil {
+		return fmt.Errorf("failure to get all keys from the blockstore: %w", err)
+	}
+	for proofCid := range proofs {
+		node, err := w.store.Get(ctx, proofCid)
+		if err != nil {
+			return fmt.Errorf("failure to get proof from the blockstore: %w", err)
+		}
+		// we chop off the first byte, as it is an unnecessary type byte.
+		cid, err := plugin.CidFromNamespacedSha256(nmt.Sha256Namespace8FlaggedInner(node.RawData()[1:]))
+		if err != nil {
+			return fmt.Errorf("failure to get cid: %w", err)
+		}
+		err = util.LdWrite(w.w, cid.Bytes(), node.RawData())
```

Comment:
Actually, we can start chopping off the additional namespaces bytes from the beginning because this optimization will require us to have special ugly ifs or migration logic later on.

Reply:
The namespace that NMTWrapper appends. You trim it on reading side https://github.com/celestiaorg/celestia-node/pull/1158/files#diff-c6030b81965c85f0d3f6416403aba04a641d95bce1a490b57a3d9c01caacf111R239, so we can avoid storing it

Reply:
Can't do this unfortunately because the nmt hasher requires it twice (see HashLeaf in hasher.go), so the CarReader isn't even able to read the next node from the file

Reply:
So, I think we need to wait until it is actually fully removed to update this (it will be obvious anyways, because then the tests will fail)

Reply:
> I just don't wanna store something that does need to be stored. Also, avoid such optimizations as Vlad is currently doing, where we trim a byte because we didn't trim from the beginning. Further, we won't have the privilege of ignoring backward compatibility. We don't have to do this in your PR, but we can do this during the epic.

An issue in the epic for this https://github.com/celestiaorg/celestia-node/issues/1233